tools/code_analyzer/op_dependency.cpp

```diff
@@ -0,0 +1,681 @@
+  // [PATH][10][INST] %18 = alloca %"struct.c10::OperatorName", align 8
+  // [PATH][11][INST] invoke void @c10::Dispatcher::findSchema(c10::OperatorName
+  //                  const&)(%"class.c10::optional.105"* nonnull sret %17,
+  //                  %"class.c10::Dispatcher.6320"* nonnull %45,
+  //                  %"struct.c10::OperatorName"* nonnull dereferenceable(16)
+  //                  %18)
+  static void scanOpInvocation(
+      VALUE_SET& instructions, SET* opSchemaStrs, GRAPH* functionToSchemaStrs) {
+    for (auto V : instructions) {
+      auto I = dyn_cast<Instruction>(V);
+      // We only need to process call/invoke instructions.
+      if (!I || !CallSite(I)) continue;
+      if (Verbose > 2) {
+        std::cerr << "[DEBUG][CALL][INST] " << *I << std::endl;
+      }
+      std::string caller = I->getFunction()->getName();
+      SET visitedOps;
+      scanOpSchemaStrAndFunction(I, {}, &visitedOps, nullptr);
+      if (visitedOps.size() != 1) {
+        std::cerr << "[WARNING] found " << visitedOps.size() << " ops ( ";
+        for (auto& op : visitedOps) {
+          std::cerr << op << " ";
+        }
+        std::cerr << ") in a invocation call in function: "
+                  << demangle(caller) << std::endl;
+      }
+      for (const auto& op : visitedOps) {
+        opSchemaStrs->insert(op);
+        (*functionToSchemaStrs)[caller].insert(op);
+        if (Verbose) {
+          std::cerr << "[DEBUG][OP_CALL] " << demangle(caller) << " => "
+                    << op << std::endl;
+        }
+      }
+    }
+  }
+
+  static void extractStringValue(
+      Value* V, const std::function<void(const std::string&)>& CB) {
+    if (auto array = dyn_cast<ConstantDataArray>(V)) {
+      // Normal case for c-style string literal and "std::basic_string".
+      if (array->isCString()) {
+        CB(array->getAsCString().str());
+      } else if (array->isString()) {
+        std::cerr << "[WARNING] ignore non-C string: "
+                  << array->getAsString().str() << std::endl;
+      }
+    } else if (auto CI = dyn_cast<ConstantInt>(V)) {
+      // Short string literal might be encoded into constant integer, e.g.:
+      // "aten::AA" => 4702103508586165345 (0x41413A3A6E657461)
+      // This can be tricky as it depends on consistent endianness/size.
+      // Seen this case for "std::__1::basic_string" ABI.
+      int64_t intValue = CI->getZExtValue();
+      auto data = reinterpret_cast<const char*>(&intValue);
+      CB({data, data + sizeof(int64_t)/sizeof(char)});
+    } else if (auto C = dyn_cast<Constant>(V)) {
+      // Short string literal might be in a constant vector, e.g.:
+      // store <2 x i64> <i64 8, i64 4702103508586165345>, <2 x i64>* %25
+      // Recursively extract each element to cover this case.
+      // Seen this case for "std::__cxx11::basic_string" ABI.
+      for (unsigned i = 0; auto elem = C->getAggregateElement(i); ++i) {
```

Comment:
See my reply above - in case 3 the encoded string is part of the ConstantAggregate.

Reply:
I see, but my question is still relevant: in the condition you're only guaranteeing that `C` is `llvm::Constant`, which does not have to be `llvm::ConstantAggregate`.

Reply:
Yeah, I can change the if condition to `dyn_cast<ConstantAggregate>` but at the time of writing this I wasn't 100% sure there was no other Constant subtype that implements `getAggregateElement`. Since Constant has this method so I didn't bother checking (why don't they move this method to ConstantAggregate? :>)
Ok, I think I can change this to ConstantAggregate - if we found exception in future we can add additional logic here (the current logic probably won't handle unknown new cases correctly anyway).

Reply:
Hmm, just tested, it actually derives from `llvm::ConstantData` instead of `llvm::ConstantAggregate`:
https://llvm.org/doxygen/classllvm_1_1Constant.html

Reply:
I guess I will go as is for now - the philosophy of this pass is to optimize for recall instead of precision (that's why we search for all uses and operands of all relevant instructions). It's kinda similar here, for whatever reason if a constant has non-null aggregate element that happens to look like an op schema (starts with "aten::") then it's probably not a false positive.
Inside `scanOpRegistration` and `scanOpInvocation` we count number of schema strings we find per API call and print warnings if it's not 1-1 mapping, so I feel we should know when it's broken.

Reply:
The reason it is in `Constant` is probably because it also handles undef. I've just checked the source and it seems it handles non-aggregate constants automatically, so keeping `Constant` should be fine: https://github.com/llvm/llvm-project/blob/master/llvm/lib/IR/Constants.cpp#L359-L373